nomad/volumewatcher/volume_watcher.go

```diff
@@ -0,0 +1,382 @@
+package volumewatcher
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	log "github.com/hashicorp/go-hclog"
+	memdb "github.com/hashicorp/go-memdb"
+	multierror "github.com/hashicorp/go-multierror"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/nomad/state"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// volumeWatcher is used to watch a single volume and trigger the
+// scheduler when allocation health transitions.
+type volumeWatcher struct {
+	// v is the volume being watched
+	v *structs.CSIVolume
+
+	// state is the state that is watched for state changes.
+	state *state.StateStore
+
+	// updateClaims is the function used to apply claims to raft
+	updateClaims updateClaimsFn
+
+	// server interface for CSI client RPCs
+	rpc ClientRPC
+
+	logger      log.Logger
+	shutdownCtx context.Context // parent context
+	ctx         context.Context // own context
+	exitFn      context.CancelFunc
+
+	// updateCh is triggered when there is an updated volume
+	updateCh chan *structs.CSIVolume
+
+	wLock   sync.RWMutex
+	running bool
+}
+
+// newVolumeWatcher returns a volume watcher that is used to watch
+// volumes
+func newVolumeWatcher(parent *Watcher, vol *structs.CSIVolume) *volumeWatcher {
+
+	w := &volumeWatcher{
+		updateCh:     make(chan *structs.CSIVolume, 1),
+		updateClaims: parent.updateClaims,
+		v:            vol,
+		state:        parent.state,
+		rpc:          parent.rpc,
+		logger:       parent.logger.With("volume_id", vol.ID, "namespace", vol.Namespace),
+		shutdownCtx:  parent.ctx,
+	}
+
+	// Start the long lived watcher that scans for allocation updates
+	w.Start()
```

Comment:
For the long term, I'm a little worried about starting a go routine for every volume. It seems at least possible that some operators could set us up for more goroutines than we want running for these, maybe in a followup if warranted we could use a fixed pool?

Reply:
I think that's a great idea. But it occurs to me we could also simply `Stop()` the goroutine when we've completed processing an incoming update (when we've determined either there are no more past claims to process or no more claims at all). This is a small change but will require some test reworking so I'm going to spin it off into https://github.com/hashicorp/nomad/issues/7837